skyportal/handlers/api/user.py

```diff
@@ -27,11 +28,87 @@ def get(self, user_id=None):
               application/json:
                 schema: Error
         """
-        user = User.query.options(joinedload(User.acls)).get(int(user_id))
+        user = User.query.get(int(user_id))
         if user is None:
             return self.error(f'Invalid user ID ({user_id}).')
         else:
-            return self.success(data=user)
+            user_info = user.to_dict()
+            user_info['acls'] = user.acls
+            return self.success(data=user_info)
+
+    @permissions(["Manage users"])
+    def post(self):
+        """
+        ---
+        description: Add a new user
+        requestBody:
+          content:
+            application/json:
+              schema:
+                type: object
+                properties:
+                  username:
+                    type: string
+                    description: User's email address
+                  roles:
+                    type: array
+                    items:
+                      type: string
+                    enum: {list(role_acls)}
+                    description: |
+                      List of user roles. Defaults to `[Full user]`. Will be overridden
+                      by `groupIDsAndAdmin` on a per-group basis.
+                  groupIDsAndAdmin:
+                    type: array
+                    items:
+                      type: array
+                    description: |
+                      Array of 2-element arrays `[groupID, admin]` where `groupID`
+                      is the ID of a group that the new user will be added to and
+                      `admin` is a boolean indicating whether they will be an admin in
+                      that group, e.g. `[[group_id_1, true], [group_id_2, false]]`
+                required:
+                  - username
+        responses:
+          200:
+            content:
+              application/json:
+                schema:
+                  allOf:
+                    - $ref: '#/components/schemas/Success'
+                    - type: object
+                      properties:
+                        data:
+                          type: object
+                          properties:
+                            id:
+                              type: integer
+                              description: New user ID
+        """
+        data = self.get_json()
+        roles = data.get("roles", ["Full user"])
+        group_ids_and_admin = data.get("groupIDsAndAdmin", [])
+
+        # Add user
+        user = User(username=data["username"], role_ids=roles)
```

Comment:
We should be normalizing the `username` here (and potentially elsewhere) to insure we have all lowercase, no escaped characters, allowed email formats, etc. We certainly dont want usernames like Josh!😂.

A similar checking and normalization could be done for `group.name` upon creation. Ie. only allow alphanumeric characters and underscores.

Reply:
Done